### DISCUSSION TITLE
Help with frontstage

### DISCUSSION THREAD

**Discussion by VedankNaik** [iTwin/itwinjs-core#5369] [Q&A] [answered]
Hello, 

I am developing an application that allows toggling between single or multiple views using fronstages. Initially, I have an iModel in the viewer with MarkerPinDecorator and RealityData. When switching to two views, I want viewer 1 to display the initial iModel with MarkerPinDecorator and RealityData, while viewer 2 will have custom content.
While I was able to achieve switching between one or two views and view the iModel in the viewer 1, the MarkerPinDecorator and RealityData disappear after switching. I want them to persist in the viewer even after toggling between views. How can I fix this?

Thanks.

**Comment by NancyMcCallB** [comment]
Hi @VedankNaik,

Are you activating a new frontstage or just changing the content layout to include two views? Because the views reinitialize, you may have to reapply the reality data and marker display after changing.

Nancy McCall

**Reply by VedankNaik** [reply]
Hi @NancyMcCallB,

I'm creating new content group and setting it as active group each time. Can this be achieved without reapply?

Thanks.

**Reply by NancyMcCallB** [reply]
I have not tried to do that in the past. My gut feeling is no, but I'll run some experiments and get back to you.

**Reply by NancyMcCallB** [reply]
@VedankNaik , can you copy in the part of your code that defines and activates the new content group?

**Reply by VedankNaik** [reply]
```
export const FrontStageChange = async (
  checked: boolean,
  iModel: IModelConnection
) => {
  const viewCreator = new ViewCreator3d(iModel);
  const options: ViewCreator3dOptions = {
    cameraOn: true,
    skyboxOn: false,
    standardViewId: 6,
    useSeedView: false,
    vpAspect: 1,
  };
  const viewState3d = await viewCreator.createDefaultView(options);
  const activeFrontstageDef = FrontstageManager.activeFrontstageDef;
  if (activeFrontstageDef) {
    const contentPropsArray: ContentProps[] = [];
    let contentGroupProps: ContentGroupProps;
    if (checked) {
      contentPropsArray.push({
        id: "3dViewportControl",
        classId: IModelViewportControl,
        applicationData: {
          viewState: viewState3d,
          iModelConnection: iModel,
        },
      });
      contentPropsArray.push({
        id: "CustomControl",
        classId: CustomContent,
      });
      contentGroupProps = {
        id: "MultiGroup",
        layout: {
          id: "TwoHalvesHorizontal",
          verticalSplit: {
            id: "TwoHalvesHorizontalSplit",
            percentage: 0.5,
            left: 0,
            right: 1,
          },
        },
        contents: contentPropsArray,
      };
    } else {
      contentPropsArray.push({
        id: "3dViewportControl",
        classId: IModelViewportControl,
        applicationData: {
          viewState: viewState3d,
          iModelConnection: iModel,
        },
      });
      contentGroupProps = {
        id: "SingleGroup",
        layout: StandardContentLayouts.singleView,
        contents: contentPropsArray,
      };
    }

    if (activeFrontstageDef.contentGroupProvider)
      contentGroupProps =
        activeFrontstageDef.contentGroupProvider.applyUpdatesToSavedProps(
          contentGroupProps
        );

    const contentGroup = new ContentGroup(contentGroupProps);
    await FrontstageManager.setActiveContentGroup(contentGroup);
  }
};

```

**Reply by NancyMcCallB** [reply]
Looks like you're creating the default view each time. When you switch to the new layout, have your tried cloning the open viewport and using that? Something like:

        const vp = activeFrontstageDef.contentControls[0].viewport;
        .
        .
       applicationData:
             {
               viewState: vp.view.clone(),
               iModelConnection: vp.view.iModel,


Or perhaps even reusing the viewport directly in the content view you don't want to change?

**Reply by VedankNaik** [reply]
Cloning the viewport fixed the reality data, but markers (decorators) are still missing.

**Reply by NancyMcCallB** [reply, marked as answer]
Hi, @VedankNaik ,

Sorry for the delay -- I don't know that much about the decorators. I think that after you clone your viewport, you should replace the viewport in your [MarkerSet ](https://www.itwinjs.org/reference/core-frontend/views/markerset/) -- something like myMarkerSet.changeViewport(clonedVp); It should then display your markers in the new viewport.

Many thanks to @Josh-Schifter and his [Marker Pins](https://www.itwinjs.org/sandboxes/iTwinPlatform/Marker%20Pins) sandbox.

**Reply by VedankNaik** [reply]
Thanks @NancyMcCallB. The issue has been fixed.
